File: Server/app/device_auth.py
"""Device proofs using the SCRAM-SHA-256 construction (RFC 5802/7677).

The application exchanges structured WebSocket messages, not SASL frames.
Passwords retain their literal UTF-8 encoding, without SASLprep normalization.
Only salt, iterations, StoredKey and ServerKey are persisted.
"""

import base64
from dataclasses import dataclass
import hashlib
import hmac
import secrets
import time

ITERATIONS = 600_000
CHALLENGE_TTL_SECONDS = 60


def encode(value: bytes) -> str:
    return base64.b64encode(value).decode("ascii")


def credentials(password: str) -> dict:
    salt = secrets.token_bytes(16)
    salted = hashlib.pbkdf2_hmac("sha256", password.encode("utf-8"), salt, ITERATIONS)
    client_key = hmac.digest(salted, b"Client Key", "sha256")
    return {
        "access_auth_salt": encode(salt),
        "access_auth_iterations": ITERATIONS,
        "access_auth_stored_key": encode(hashlib.sha256(client_key).digest()),
        "access_auth_server_key": encode(hmac.digest(salted, b"Server Key", "sha256")),
    }


def auth_message(device_id, client_nonce, nonce, salt, iterations) -> bytes:
    # SCRAM's username escaping keeps the transcript unambiguous.
    name = device_id.replace("=", "=3D").replace(",", "=2C")
    return (
        f"n={name},r={client_nonce},r={nonce},s={salt},i={iterations},c=biws,r={nonce}"
    ).encode("utf-8")


def verify_proof(
    stored_key: str, server_key: str, transcript: bytes, proof: str
) -> str | None:
    try:
        stored = base64.b64decode(stored_key, validate=True)
        server = base64.b64decode(server_key, validate=True)
        supplied = base64.b64decode(proof, validate=True)
        if not all(len(value) == 32 for value in (stored, server, supplied)):
            return None
    except ValueError:
        return None
    signature = hmac.digest(stored, transcript, "sha256")
    client_key = bytes(a ^ b for a, b in zip(supplied, signature))
    if not hmac.compare_digest(hashlib.sha256(client_key).digest(), stored):
        return None
    return encode(hmac.digest(server, transcript, "sha256"))


@dataclass(frozen=True)
class Challenge:
    id: str
    device_id: str
    client_nonce: str
    nonce: str
    salt: str
    iterations: int
    expires_at: float

    @property
    def transcript(self):
        return auth_message(
            self.device_id, self.client_nonce, self.nonce, self.salt, self.iterations
        )

    def response(self):
        return {
            "type": "session_challenge",
            "deviceId": self.device_id,
            "challengeId": self.id,
            "clientNonce": self.client_nonce,
            "nonce": self.nonce,
            "salt": self.salt,
            "iterations": self.iterations,
            "expiresIn": CHALLENGE_TTL_SECONDS,
        }


class DeviceChallenges:
    """Owned by one authenticated console connection; never shared or persisted."""

    def __init__(self, limit):
        self.limit = limit
        self.pending: dict[str, Challenge] = {}

    def issue(self, device, client_nonce):
        now = time.monotonic()
        self.pending = {
            key: value for key, value in self.pending.items() if value.expires_at > now
        }
        if device.id not in self.pending and len(self.pending) >= self.limit:
            return None
        challenge = Challenge(
            secrets.token_urlsafe(32),
            device.id,
            client_nonce,
            client_nonce + secrets.token_hex(32),
            device.access_auth_salt,
            device.access_auth_iterations,
            now + CHALLENGE_TTL_SECONDS,
        )
        self.pending[device.id] = challenge
        return challenge

    def consume(self, device_id, challenge_id):
        # Consume before validation, with no await: even a failed guess is single-use.
        challenge = self.pending.pop(device_id, None)
        if (
            challenge
            and challenge.id == challenge_id
            and challenge.expires_at > time.monotonic()
        ):
            return challenge
        return None

File: Server/app/static/device-auth.js
// SCRAM-SHA-256 proof construction, with literal UTF-8 passwords (no SASLprep).
const DeviceAuth = (() => {
  const encode = value => btoa(String.fromCharCode(...value));
  const decode = value => Uint8Array.from(atob(value), char => char.charCodeAt(0));
  const utf8 = value => new TextEncoder().encode(value);
  async function hmac(key, message) {
    const imported = await crypto.subtle.importKey("raw", key,
      {name: "HMAC", hash: "SHA-256"}, false, ["sign"]);
    return new Uint8Array(await crypto.subtle.sign("HMAC", imported, message));
  }
  async function proof(key, challenge) {
    const {deviceId, clientNonce, nonce, salt, iterations} = challenge;
    const saltBytes = decode(salt);
    if (saltBytes.length !== 16 || !Number.isInteger(iterations) || iterations < 4096 || iterations > 1000000) {
      throw new Error("Parametri di autenticazione non validi.");
    }
    const name = deviceId.replace(/=/g, "=3D").replace(/,/g, "=2C");
    const transcript = utf8(`n=${name},r=${clientNonce},r=${nonce},s=${salt},i=${iterations},c=biws,r=${nonce}`);
    const salted = new Uint8Array(await crypto.subtle.deriveBits(
      {name: "PBKDF2", hash: "SHA-256", salt: saltBytes, iterations}, key, 256));
    let clientKey, serverKey;
    try {
      clientKey = await hmac(salted, utf8("Client Key"));
      serverKey = await hmac(salted, utf8("Server Key"));
      const storedKey = new Uint8Array(await crypto.subtle.digest("SHA-256", clientKey));
      const signature = await hmac(storedKey, transcript);
      return {
        proof: encode(clientKey.map((byte, index) => byte ^ signature[index])),
        serverProof: encode(await hmac(serverKey, transcript)),
      };
    } finally {
      salted.fill(0);
      clientKey?.fill(0);
      serverKey?.fill(0);
    }
  }
  return {
    nonce: () => Array.from(crypto.getRandomValues(new Uint8Array(32)), byte => byte.toString(16).padStart(2, "0")).join(""),
    importPassword: bytes => crypto.subtle.importKey("raw", bytes, "PBKDF2", false, ["deriveBits"]),
    proof,
  };
})();
